response.go

```diff
@@ -35,78 +36,75 @@ type Response struct {
 }
 
 func (or *Response) readSGMLHeaders(r *bufio.Reader) error {
-	var seenHeader, seenVersion bool = false, false
-	for {
-		// Some financial institutions do not properly leave an empty line after the last header.
-		// Avoid attempting to read another header in that case.
-		next, err := r.Peek(1)
-		if err != nil {
-			return err
-		}
-		if next[0] == '<' {
-			break
-		}
+	b, err := r.ReadSlice('<')
+	if err != nil {
+		return err
+	}
 
-		line, err := r.ReadString('\n')
-		if err != nil {
-			return err
-		}
-		// r.ReadString leaves the '\n' on the end...
-		line = strings.TrimSpace(line)
-
-		if len(line) == 0 {
-			if seenHeader {
-				break
-			} else {
-				continue
-			}
-		}
-		header := strings.SplitN(line, ":", 2)
-		if header == nil || len(header) != 2 {
-			return errors.New("OFX headers malformed")
-		}
+	s := string(b)
+	err = r.UnreadByte()
+	if err != nil {
+		return err
+	}
 
-		// Some OFX servers put a space after the colon
-		headervalue := strings.TrimSpace(header[1])
+	// According to the latest OFX SGML spec (1.6), headers should be CRLF-separated
+	// and written as KEY:VALUE. However, some banks include a whitespace after the
+	// colon (KEY: VALUE), while others include no line breaks at all. The spec doesn't
+	// require a line break after the OFX headers, but it is allowed, and will be
+	// optionally captured & discarded by the trailing `\s*`. Valid SGML headers must
+	// always be present in exactly this order, so a regular expression is acceptable.
+	headerExp := regexp.MustCompile(
+		`OFXHEADER:\s*(?P<OFXHEADER>\d+)\s*` +
+			`DATA:\s*(?P<DATA>[A-Z]+)\s*` +
+			`VERSION:\s*(?P<VERSION>\d+)\s*` +
+			`SECURITY:\s*(?P<SECURITY>[\w]+)\s*` +
+			`ENCODING:\s*(?P<ENCODING>[A-Z0-9-]+)\s*` +
+			`CHARSET:\s*(?P<CHARSET>[\w-]+)\s*` +
+			`COMPRESSION:\s*(?P<COMPRESSION>[A-Z]+)\s*` +
+			`OLDFILEUID:\s*(?P<OLDFILEUID>[\w-]+)\s*` +
+			`NEWFILEUID:\s*(?P<NEWFILEUID>[\w-]+)\s*`)
+
+	matches := headerExp.FindStringSubmatch(s)
```

Comment:
Because `FindStringSubmatch` allows matching a partial string, I think it's possible for us to leave some unparsed text. Arguably it's malformatted OFX if there's extra 'stuff' after the header and before the first '<', but I wouldn't put it past them. Would this still work correctly if we prefix the regular expression with '^' and suffix it with '$' to guard against this?

Reply:
Good call, that shouldn't break matching and should catch other weird malformed headers 👍

Reply:
Hmm, adding the leading `^` match is causing test failures since a test input (`invalidResponse` in `response_test.go:178`) has a leading newline. It's resolved by changing the first match to allow optional leading whitespace; what do you think?

```diff
- 		`^OFXHEADER:\s*(?P<OFXHEADER>\d+)\s*` +
+ 		`^\s*OFXHEADER:\s*(?P<OFXHEADER>\d+)\s*` +
```

Reply:
Interesting. Is that single failing test the only issue? If so, we can also just fix that test - I think the leading newline there is accidental and can be removed without sacrificing the intent of the test.

That said, how pernicious can a little leading whitespace be? :stuck_out_tongue_closed_eyes:

Reply:
Yeah, looks like that one's the only failing test to me. I'd say if omitting the leading whitespace from the regular expression works for what you need, let's just remove the newline from the test and leave this in the more strict form for now.